File: libs/python/stonecharts/charts/flame_chart.py
"""Flame chart (time-ordered) renderer: ChartSpec -> SVG string.

Per-thread stack frames over wall-clock time, drawn as horizontal floating
bars at [start, end] per depth level. Depth 0 = root at bottom, increasing
upward. Rides the shared cartesian frame with orientation="horizontal",
x_scale="band", include_zero=False.
"""

from __future__ import annotations

from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, render_cartesian


def render_svg(spec) -> str:
    max_depth = 0
    all_times: list[float] = []
    for s in spec.series:
        for fr in s.frames or []:
            if fr.depth > max_depth:
                max_depth = fr.depth
            all_times.append(fr.x)
            all_times.append(fr.x2)

    depth_cats = [str(d) for d in range(max_depth + 1)]
    spec.x_axis.categories = depth_cats

    for s in spec.series:
        while len(s.data) < len(depth_cats):
            s.data.append(0.0)

    if spec.y_axis.min is None:
        spec.y_axis.min = spec.x_axis.min if spec.x_axis.min is not None else (min(all_times) if all_times else 0.0)
    if spec.y_axis.max is None:
        spec.y_axis.max = spec.x_axis.max if spec.x_axis.max is not None else (max(all_times) if all_times else 0.0)

    return render_cartesian(spec, "Flame chart", "band", _flame_marks, include_zero=False, orientation="horizontal")


PAD = 0.2
LABEL_MIN_PX = 40
CHAR_WIDTH = 6.5


def _flame_marks(fr: CartesianFrame, p: list[str]) -> None:
    if fr.n <= 0:
        return

    lane_height = fr.plot_h / fr.n
    bar_thickness = lane_height * (1 - PAD)

    def ypix_band(depth: int) -> float:
        inverted = fr.n - 1 - depth
        return fr.plot_y + lane_height * inverted + lane_height / 2

    def xval(v: float) -> float:
        return fr.value_pix(v)

    for si, s in enumerate(fr.spec.series):
        st = fr.styles[si]
        fill = st.fill

        p.append(f'<g class="sc-series" data-series="{si}">')

        for frame in s.frames or []:
            start = frame.x
            end = frame.x2
            depth = frame.depth
            name = frame.name or ""

            frame_fill = fill
            if frame.color:
                frame_fill = frame.color

            x_left = xval(min(start, end))
            x_right = xval(max(start, end))
            w = x_right - x_left
            if w < 1.0:
                w = 1.0
            cy = ypix_band(depth)
            cx = xval((start + end) / 2)
            top = cy - bar_thickness / 2

            duration = end - start
            depth_label = str(depth)

            data_attrs = (
                f'class="sc-frame sc-point" '
                f'data-series="{si}" data-series-name="{esc(s.name)}" '
                f'data-x="{esc(depth_label)}" data-y="{esc(fmt_num(start))}" '
                f'data-start="{esc(fmt_num(start))}" data-end="{esc(fmt_num(end))}" '
                f'data-depth="{depth}" data-name="{esc(name)}" '
                f'data-duration="{esc(fmt_num(duration))}" '
                f'data-color="{esc(frame_fill)}" data-r="{fmt_num(3.5)}" data-r-hover="{fmt_num(6)}" '
                f'cx="{cx:.1f}" cy="{cy:.1f}"'
            )

            p.append(
                f"<rect {data_attrs} "
                f'x="{x_left:.1f}" y="{top:.1f}" width="{w:.1f}" height="{bar_thickness:.1f}" fill="{esc(frame_fill)}"/>'
            )

            if name and w >= LABEL_MIN_PX:
                max_chars = int(w / CHAR_WIDTH)
                label = name
                if len(label) > max_chars:
                    label = label[: max(max_chars - 1, 0)] + "…"
                label_y = cy + 3.5
                p.append(
                    f'<text class="sc-frame-label" x="{cx:.1f}" y="{label_y:.1f}" '
                    f'text-anchor="middle" font-size="10" fill="#ffffff">{esc(label)}</text>'
                )

        p.append("</g>")

File: libs/go/flame_chart.go
package stonecharts

import (
	"fmt"
	"strings"
)

func renderFlameChartSVG(spec *ChartSpec) string {
	mod := *spec
	mod.Series = make([]Series, len(spec.Series))
	copy(mod.Series, spec.Series)
	for i := range mod.Series {
		d := make([]float64, len(spec.Series[i].Data))
		copy(d, spec.Series[i].Data)
		mod.Series[i].Data = d
	}

	maxDepth := 0
	var allTimes []float64
	for _, s := range mod.Series {
		for _, fr := range s.Frames {
			if fr.Depth > maxDepth {
				maxDepth = fr.Depth
			}
			allTimes = append(allTimes, fr.X, fr.X2)
		}
	}

	depthCats := make([]string, maxDepth+1)
	for d := 0; d <= maxDepth; d++ {
		depthCats[d] = fmt.Sprintf("%d", d)
	}
	mod.XAxis.Categories = depthCats

	for i := range mod.Series {
		for len(mod.Series[i].Data) < len(depthCats) {
			mod.Series[i].Data = append(mod.Series[i].Data, 0.0)
		}
	}

	if mod.YAxis.Min == nil {
		v := 0.0
		if mod.XAxis.Min != nil {
			v = *mod.XAxis.Min
		} else if len(allTimes) > 0 {
			v = allTimes[0]
			for _, t := range allTimes[1:] {
				if t < v {
					v = t
				}
			}
		}
		mod.YAxis.Min = &v
	}
	if mod.YAxis.Max == nil {
		v := 0.0
		if mod.XAxis.Max != nil {
			v = *mod.XAxis.Max
		} else if len(allTimes) > 0 {
			v = allTimes[0]
			for _, t := range allTimes[1:] {
				if t > v {
					v = t
				}
			}
		}
		mod.YAxis.Max = &v
	}

	return renderCartesian(&mod, "Flame chart", "band", flameMarks, false, "horizontal")
}

const (
	flamePad       = 0.2
	labelMinPx     = 40.0
	flameCharWidth = 6.5
)

func flameMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}

	laneHeight := f.plotH / float64(f.n)
	barThickness := laneHeight * (1 - flamePad)

	ypixBand := func(depth int) float64 {
		inverted := f.n - 1 - depth
		return f.plotY + laneHeight*float64(inverted) + laneHeight/2
	}

	xval := func(v float64) float64 {
		return f.valuePix(v)
	}

	for si, s := range f.spec.Series {
		st := f.styles[si]
		fill := st.fill

		p.WriteString(fmt.Sprintf(`<g class="sc-series" data-series="%d">`, si))

		for _, frame := range s.Frames {
			start := frame.X
			end := frame.X2
			depth := frame.Depth
			name := frame.Name

			frameFill := fill
			if frame.Color != "" {
				frameFill = frame.Color
			}

			xLeft := xval(start)
			xRight := xval(end)
			if start > end {
				xLeft, xRight = xRight, xLeft
			}
			w := xRight - xLeft
			if w < 1.0 {
				w = 1.0
			}
			cy := ypixBand(depth)
			cx := xval((start + end) / 2)
			top := cy - barThickness/2

			duration := end - start
			depthLabel := fmt.Sprintf("%d", depth)

			dataAttrs := fmt.Sprintf(
				`class="sc-frame sc-point" data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-start="%s" data-end="%s" data-depth="%d" data-name="%s" data-duration="%s" data-color="%s" data-r="%s" data-r-hover="%s" cx="%s" cy="%s"`,
				si, esc(s.Name), esc(depthLabel), esc(fmtNum(start)), esc(fmtNum(start)), esc(fmtNum(end)),
				depth, esc(name), esc(fmtNum(duration)), esc(frameFill), fmtNum(3.5), fmtNum(6),
				f1(cx), f1(cy))

			p.WriteString(fmt.Sprintf(
				`<rect %s x="%s" y="%s" width="%s" height="%s" fill="%s"/>`,
				dataAttrs, f1(xLeft), f1(top), f1(w), f1(barThickness), esc(frameFill)))

			if name != "" && w >= labelMinPx {
				maxChars := int(w / flameCharWidth)
				label := name
				if len(label) > maxChars {
					if maxChars-1 > 0 {
						label = label[:maxChars-1] + "…"
					} else {
						label = "…"
					}
				}
				labelY := cy + 3.5
				p.WriteString(fmt.Sprintf(
					`<text class="sc-frame-label" x="%s" y="%s" text-anchor="middle" font-size="10" fill="#ffffff">%s</text>`,
					f1(cx), f1(labelY), esc(label)))
			}
		}

		p.WriteString(`</g>`)
	}
}

func flameChartDataTable(spec *ChartSpec) string {
	var b strings.Builder
	b.WriteString(`<table class="sc-visually-hidden">`)
	if spec.Title != "" {
		b.WriteString("<caption>" + esc(spec.Title) + "</caption>")
	}
	b.WriteString(`<thead><tr><th scope="col">Series</th><th scope="col">Depth</th>` +
		`<th scope="col">Start</th><th scope="col">End</th>` +
		`<th scope="col">Duration</th><th scope="col">Name</th></tr></thead><tbody>`)
	for _, s := range spec.Series {
		for _, fr := range s.Frames {
			duration := fr.X2 - fr.X
			name := fr.Name
			b.WriteString(`<tr><th scope="row">` + esc(s.Name) + `</th><td>` +
				fmt.Sprintf("%d", fr.Depth) + `</td><td>` +
				esc(fmtNum(fr.X)) + `</td><td>` + esc(fmtNum(fr.X2)) + `</td><td>` +
				esc(fmtNum(duration)) + `</td><td>` + esc(name) + `</td></tr>`)
		}
	}
	b.WriteString("</tbody></table>")
	return b.String()
}
